src/address.c

```diff
@@ -0,0 +1,149 @@
+        unsigned char redeem_script[HASH160_LEN + 2];
+        redeem_script[0] = OP_0;
+        redeem_script[1] = HASH160_LEN;
+
+        if (wally_hash160(hdkey->pub_key, sizeof(hdkey->pub_key), redeem_script + 2, HASH160_LEN) != WALLY_OK)
+            return WALLY_EINVAL;
+
+        // P2SH address = version (e.g. 0x05) + ripemd160(sha256(redeem_script))
+        address[0] = (unsigned char) version & 0xff;
+        if (wally_hash160(redeem_script, sizeof(redeem_script), address + 1, HASH160_LEN) != WALLY_OK)
+            return WALLY_EINVAL;
+    }
+
+    ret = wally_base58_from_bytes(address, sizeof(address), BASE58_FLAG_CHECKSUM, output);
+
+    wally_clear(address, sizeof(address));
+    return ret;
+}
+
+int wally_bip32_key_to_addr_segwit(const struct ext_key *hdkey, const char *addr_family,
+                                   uint32_t flags, char **output)
+{
+    int ret;
+
+    // Witness program bytes, including the version and data push opcode.
+    unsigned char witness_program_bytes[HASH160_LEN + 2];
+    witness_program_bytes[0] = OP_0;
+    witness_program_bytes[1] = HASH160_LEN;
+
+    if (wally_hash160(hdkey->pub_key, sizeof(hdkey->pub_key), witness_program_bytes + 2, HASH160_LEN) != WALLY_OK)
+        return WALLY_EINVAL;
+
+    ret = wally_addr_segwit_from_bytes(witness_program_bytes, HASH160_LEN + 2, addr_family, flags, output);
+    return ret;
+}
+
+int wally_address_to_scriptpubkey(const char *addr, uint32_t flags, unsigned char *bytes_out,
+                                  size_t len, size_t *written)
+{
+    int ret;
+
+    uint32_t version;
+    unsigned char bytes_base58_decode[1 + HASH160_LEN + BASE58_CHECKSUM_LEN];
+    size_t written_base58_decode;
+
+    if (written)
+        *written = 0;
+
+    if (!(flags == WALLY_NETWORK_BITCOIN_MAINNET || flags == WALLY_NETWORK_BITCOIN_TESTNET))
+        return WALLY_EINVAL;
+
+    // Check minimum length of bytes_out:
+    // P2PKH 25 bytes: OP_DUP OP_HASH160 [pub_key_hash] OP_EQUALVERIFY OP_CHECKSIG
+    // P2SH  23 bytes: OP_HASH160 [script_hash] OP_EQUAL
+    if (len < 25)
+        return WALLY_EINVAL;
+
+    // This returns WALLY_OK even if addr is too long for bytes_base58_decode
+    if (wally_base58_to_bytes(addr, BASE58_FLAG_CHECKSUM, bytes_base58_decode, sizeof(bytes_base58_decode), &written_base58_decode) != WALLY_OK)
+        return WALLY_EINVAL;
+
+    // Belt and suspenders check; wally_base58_to_bytes with BASE58_FLAG_CHECKSUM
+    // currently already catches this case.
+    if (written_base58_decode == 0)
+        return WALLY_EINVAL;
+
+    version = bytes_base58_decode[0];
+
+    switch (version) {
+    case 0: // Bitcoin P2PKH
+        if (flags & WALLY_NETWORK_BITCOIN_TESTNET)
+            return WALLY_EINVAL;
+    case 111: // Bitcoin testnet P2PKH
+        if (version == 111 && flags & WALLY_NETWORK_BITCOIN_MAINNET)
+            return WALLY_EINVAL;
+        if (written_base58_decode != HASH160_LEN + 1)
```

Comment:
Is there a reason why this check and the identical one on line 133 could not be moved up to line 102, and replace the check for == 0?

Reply:
I reorganized that code a little bit. It's actually longer, but probably more readable and less repetitive.
1. moved the input length check to a separate switch statement
2. moved the network check to a separate switch statement